### PR TITLE
Add icons to devices/sessions

### LINKED ISSUE #3683: Add Icons to Devices/Sessions on Personal Page
We could add some logos for browsers and devices to the personal page so that the devices could be better recognized by the users 😉 

Example for Dropbox:

bildschirmfoto 2017-03-02 um 17 52 29
bildschirmfoto 2017-03-02 um 17 53 06

I remember that this was in the initial drafts of this feature but ... well 😅 

cc @jancborchardt @nextcloud/designers

### PR DESCRIPTION
**Pull request by pierlon**
Fixes #3683.

--- 

Before: 
image

After:
image

### CODE REVIEW

**Comment by MorrisJobke** [comment]
Hi @pierlon - Thanks for your contribution.

Just a little heads up (CI would also have warned you about this later anyways): we require a sign-off message. This basically means, that you are the originator of the code and if not have checked that the  license is compatible with our AGPL v3+. Did you made those icons yourself or did you picked it from some source? Maybe just reference in this PR from it.

Regarding the sign-off message - see our contribution guidelines on how to adjust the commit message: https://github.com/nextcloud/server/blob/master/CONTRIBUTING.md#sign-your-work

**Comment by MorrisJobke** [comment]
cc @nextcloud/designers

File: settings/js/authtoken_view.js [hunk 1]
```diff
@@ -161,6 +180,11 @@
 					} else {
 						viewData.name = nameMap[client];
 					}
+
+					// update title - for easier view
+					viewData.title = viewData.name;
+
+					viewData.icon = '/settings/img/clients/' + iconMap[client];
```
Comment (skjnldsv) [inline settings/js/authtoken_view.js]: What if iconMap[client] doesn't exists?

**Comment by skjnldsv** [comment]
Also, icons have inconsistent sizes, margin... etc

**Comment by pierlon** [comment]
Hey @MorrisJobke, thanks for the heads up; modified my commit message. Regarding the icons, here are the sources:

Icon | URL
---|---
[Globe](https://github.com/pierlon/server/blob/642bdb8db24d07835a0e20f5180e59ae3f997293/settings/img/clients/globe.svg)|https://www.iconfinder.com/icons/172473/global_globe_internet_icon
[Apple](https://github.com/pierlon/server/blob/642bdb8db24d07835a0e20f5180e59ae3f997293/settings/img/clients/apple.svg)|https://www.iconfinder.com/icons/104869/apple_icon
[Android](https://github.com/pierlon/server/blob/642bdb8db24d07835a0e20f5180e59ae3f997293/settings/img/clients/android.svg)|https://www.iconfinder.com/icons/104870/android_icon
[Chrome](https://github.com/pierlon/server/blob/642bdb8db24d07835a0e20f5180e59ae3f997293/settings/img/clients/chrome.svg)|https://www.iconfinder.com/icons/401267/browser_chrome_icon
[Edge](https://github.com/pierlon/server/blob/642bdb8db24d07835a0e20f5180e59ae3f997293/settings/img/clients/edge.svg)|https://www.iconfinder.com/icons/843779/browser_edge_explorer_microsoft_icon
[Firefox](https://github.com/pierlon/server/blob/642bdb8db24d07835a0e20f5180e59ae3f997293/settings/img/clients/firefox.svg)|https://www.iconfinder.com/icons/104886/browser_firefox_icon
[Internet Explorer](https://github.com/pierlon/server/blob/642bdb8db24d07835a0e20f5180e59ae3f997293/settings/img/clients/ie.svg)|https://www.iconfinder.com/icons/104884/browser_explorer_internet_icon
[iPad](https://github.com/pierlon/server/blob/642bdb8db24d07835a0e20f5180e59ae3f997293/settings/img/clients/ipad.svg)|https://www.iconfinder.com/icons/383210/tablet_icon
[iPhone](https://github.com/pierlon/server/blob/642bdb8db24d07835a0e20f5180e59ae3f997293/settings/img/clients/iphone.svg)|https://www.iconfinder.com/icons/2639843/iphone_icon
[Safari](https://github.com/pierlon/server/blob/642bdb8db24d07835a0e20f5180e59ae3f997293/settings/img/clients/safari.svg)|https://www.iconfinder.com/icons/104875/browser_compass_safari_icon

File: settings/js/authtoken_view.js (hunk 1 above)
Comment (pierlon) [reply settings/js/authtoken_view.js]: If `iconMap[client]` doesn't exist, so does `client`.

**Comment by pierlon** [comment]
@skjnldsv Oh didn't notice, resizing them now.

**Comment by MariusBluem** [comment]
Linux and Windows are missing clients? 🤔 Globe is used as fallback, right?

**Comment by pierlon** [comment]
@MariusBluem Ah, yes forgot about that. I was going by the clients that were listed in `nameMap`. Adding those as well.

**Comment by MariusBluem** [comment]
No. There are client apps (iOS, android, Windows, macOS, Linux, ...) *and* browsers.

**Comment by pierlon** [comment]
Oh OK. Currently, the code only detects android and iOS client apps. I could add support for the other client apps, but I would need the user agent of each to do so.

**Comment by jancborchardt** [comment]
Good stuff! :) Just some details:

- [ ] As @skjnldsv also noted, the icons should have a consistent size: Ideally how they will be displayed, 16*16px squares. :)
- [ ] Could you place the »clients« folder in `core/img/` instead of settings? I prefer to have all the images in that one place (yes the icons of the popover are the only exceptions)
- [ ] Please use the icons.scss file to define classes like `.icon-client-firefox` to be used by the app, that’s how we do it centrally for icons. :)
- [ ] The globe icon we already have at https://github.com/pierlon/server/blob/master/core/img/places/link.svg

Thanks a lot for contributing to Nextcloud design! :) I’m adding you to the @nextcloud/designers group, and just in case you didn’t know about it, check out the design page :) https://nextcloud.com/design/

**Comment by pierlon** [comment]
@jancborchardt Thanks! Made the modifications as you had asked.

**Comment by MorrisJobke** [comment]
I looked at the licenses:

* CC-ND 3.0
* CC 3.0
* WTFPL
* "Free for commercial use"

@schiessle Is this okay for us? I guess we need to name the owners of the cc licensed icons, right? Where to do that?

**Comment by jancborchardt** [comment]
@pierlon oh and just btw, for a design/frontend pull request we usually always post a screenshot (incl before/after if relevant) so it's easier to quickly see what changed and to make it easier to review. :)

**Comment by pierlon** [comment]
@jancborchardt Updated OP with screenshots 😄

**Comment by jancborchardt** [comment]
@pierlon thanks! Also noted through that, the vertical alignment is slightly off, with the icons being a bit too far up (just ~3px).

**Comment by pierlon** [comment]
@jancborchardt Oh didn't notice, hope that's fixed now.

**Review by jancborchardt — APPROVED** [review]
Looks great now! Thanks a lot @pierlon :)

**Comment by MariusBluem** [comment]
Windows and Linux (Clients) are still missing ...

**Comment by jancborchardt** [comment]
@pierlon can you add the missing ones @MariusBluem mentioned? :)

**Comment by schiessle** [comment]
> @schiessle Is this okay for us? I guess we need to name the owners of the cc licensed icons, right? Where to do that?

> "Free for commercial use"

just to double check: "free for _commercial_ use" or "free for **non**-commercial use". The later one would be a problem.

Otherwise the licenses look fine.

I would put the information about the licenses here: https://github.com/nextcloud/server/blob/master/COPYING-README

There we already have the license of the elementary filetype icons. I would list them each with the path to the icon followed by the author and the license, something like:

```core/img/clients/android.svg by Joe Doe, licensed under <license name> (link to license if we don't ship them)````

**Comment by MorrisJobke** [comment]
> just to double check: "free for commercial use" or "free for non-commercial use". The later one would be a problem.

Yes it is free for commercial use. I will check and provide a list here.

**Comment by MorrisJobke** [comment]
@pierlon Could you add this to `COPYING-README` just before the line `Attribution information for Nextcloud ...`:

```

Licensing of some images:
* by Martin Eigtved, licensed under <CC BY 3.0> (https://creativecommons.org/licenses/by/3.0/):
   * core/img/clients/android.svg
   * core/img/clients/apple.svg
   * core/img/clients/firefox.svg
   * core/img/clients/ie.svg
   * core/img/clients/safari.svg
* by Aha-Soft, licensed under <CC BY 3.0> (https://creativecommons.org/licenses/by/3.0/):
   * core/img/clients/chrome.svg
* by Icons8, licensed under <WTFPL> (http://www.wtfpl.net/about/):
   * core/img/clients/iphone.svg
```

**Comment by MorrisJobke** [comment]
I just checked that page and the free for commercial use license doesn't require you anything and you can modify like you want.

**Comment by pierlon** [comment]
Added icons for Windows and Linux clients, and appended licenses for each to the list provided by @MorrisJobke.

Icon | URL
---|---
[Linux](https://github.com/pierlon/server/blob/eb0b8705860bae55acafb80a650969bfa15e1c04/core/img/clients/linux.svg) | [https://www.iconfinder.com/icons/1378027/circle_hovytech_linux_media_microsoft_social_system_icon](https://www.iconfinder.com/icons/1378027/circle_hovytech_linux_media_microsoft_social_system_icon)
[Windows](https://github.com/pierlon/server/blob/eb0b8705860bae55acafb80a650969bfa15e1c04/core/img/clients/windows.svg) | [https://www.iconfinder.com/icons/104799/windows_icon](https://www.iconfinder.com/icons/104799/windows_icon)

**Comment by MariusBluem** [comment]
Thanks. Windows and Linux icons were added, great. But after reviewing the code I can’t see where/when they are used in the Security section :(

**Comment by pierlon** [comment]
Forgot to mention about that. In order for those icons to correctly display, the user agent would have to match a pre-defined regex, which is not present at the moment. I'm not sure where to find user agents for those clients.

If the Windows client uses its default browser user agent, I think this would suffice:`Mozilla\/5\.0 \([^)]*Windows Phone[^)]+\)`

**Comment by pixelipo** [comment]
@MorrisJobke  did you look for the licence on the official companies' websites? Copyright claims made by icon "authors" on the iconfinder are BS, because those logos are property of Microsoft, Google, Apple, Mozilla etc. - so each of those needs to be checked separately. Here's  some of them:

https://developer.android.com/distribute/marketing-tools/brand-guidelines.html
https://design.firefox.com/photon/visuals/product-identity-assets.html
https://query.prod.cms.rt.microsoft.com/cms/api/am/binary/RE1qeRY
https://www.apple.com/legal/intellectual-property/guidelinesfor3rdparties.html

**Comment by pierlon** [comment]
@pixelipo is right. The only icons that can be used commercially without a trademark license are the Linux, Firefox and Android icons.

Brand | URL
---|---
Windows | https://query.prod.cms.rt.microsoft.com/cms/api/am/binary/RE1voQq
Edge | https://query.prod.cms.rt.microsoft.com/cms/api/am/binary/RE1q9FT
Apple & Safari | https://www.apple.com/legal/intellectual-property/guidelinesfor3rdparties.html
iPad & iPhone | https://developer.apple.com/softwarelicensing/agreements/icons.php
Internet Explorer | https://query.prod.cms.rt.microsoft.com/cms/api/am/binary/RE1qeRY
Google Chrome | https://www.google.com/intl/en/chrome/browser/privacy/eula_text.html
Firefox | https://design.firefox.com/photon/visuals/product-identity-assets.html
Android | https://developer.android.com/distribute/marketing-tools/brand-guidelines.html
Linux | http://isc.tamu.edu/~lewing/linux/

**Comment by rullzer** [comment]
So how about as a first step we just use generic icons? So just computer/phone/tablet? To avoid the legal hassles.

**Comment by pixelipo** [comment]
I agree with @rullzer - in that case we can just create our own icons and be done with it. Here's a quick 'n' dirty 5-min example by yours truly:
sample

**Comment by pierlon** [comment]
Ok made some generic icons to use. Refer to screenshots in OP to see the update if needed.

**Comment by rullzer** [comment]
Mmm this apparently doesn't work with the new auth flow. As we don't have the info then to decide. Maybe we should store this as well then.. @MorrisJobke @ChristophWurst ?

**Comment by MorrisJobke** [comment]
> Mmm this apparently doesn't work with the new auth flow. As we don't have the info then to decide. Maybe we should store this as well then.. @MorrisJobke @ChristophWurst ?

That's what I told in the beginning, but the answer was: we don't know, because the app sends the proper name of the device as info (because this is more likely to be recognised by the owner - Samsung S8 vs Android app), but I would say maybe the opposite is better, to say "Android app on phone/tablet (Vendor foo)".

**Comment by rullzer** [comment]
@MorrisJobke well we could just store both I'd say. Like I get the proper name makes sense. But having the type of device also helps...

**Comment by rullzer** [comment]
Ok I think it is best to fix this by requiring the clients to provide more info that we can properly parse. I'll look into how separatly. So we can get this in IMO then!

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Comment by MorrisJobke** [comment]
Thank you @pierlon for the nice work.



> Ok I think it is best to fix this by requiring the clients to provide more info that we can properly parse. I'll look into how separatly. So we can get this in IMO then!

Then we should adjust the code that is changed in this PR once we have more details. For now we should merge this.

**Comment by skjnldsv** [comment]
🎉

**Review by Chheomtit1993 — COMMENTED** [review]
http://www.apple.com/legal/privacy/

**Review by Chheomtit1993 — COMMENTED** [review]
http://www.apple.com/legal/privacy/

**Pull request merged**